sdk/python/cmd/pulumi-language-python/main.go

```diff
@@ -144,8 +148,365 @@ func newLanguageHost(exec, engineAddress, tracing, virtualenv string) pulumirpc.
+		createVirtualEnv = empty
+	}
+
+	// Create the virtual environment and install dependencies into it, if needed.
+	if createVirtualEnv {
+		// Make a connection to the real engine that we will log messages to.
+		conn, err := grpc.Dial(
+			host.engineAddress,
+			grpc.WithInsecure(),
+			rpcutil.GrpcChannelOptions(),
+		)
+		if err != nil {
+			return "", errors.Wrapf(err, "language host could not make connection to engine")
+		}
+
+		// Make a client around that connection.
+		engineClient := pulumirpc.NewEngineClient(conn)
+
+		// Create writers that log the output of the install operation as ephemeral messages.
+		streamID := rand.Int31() //nolint:gosec
+
+		infoWriter := &logWriter{
+			ctx:          ctx,
+			engineClient: engineClient,
+			streamID:     streamID,
+			severity:     pulumirpc.LogSeverity_INFO,
+		}
+
+		errorWriter := &logWriter{
+			ctx:          ctx,
+			engineClient: engineClient,
+			streamID:     streamID,
+			severity:     pulumirpc.LogSeverity_ERROR,
+		}
+
+		if err := python.InstallDependenciesWithWriters(
+			cwd, virtualenv, true /*showOutput*/, infoWriter, errorWriter); err != nil {
+			return "", err
+		}
+	}
+
+	// Ensure the specified virtual directory is a valid virtual environment.
+	if !python.IsVirtualEnv(virtualenv) {
+		return "", python.NewVirtualEnvError(host.virtualenv, virtualenv)
+	}
+
+	// Return the full path to the virtual environment.
+	return virtualenv, nil
+}
+
+type logWriter struct {
+	ctx          context.Context
+	engineClient pulumirpc.EngineClient
+	streamID     int32
+	severity     pulumirpc.LogSeverity
+}
+
+func (w *logWriter) Write(p []byte) (n int, err error) {
+	val := string(p)
+	if _, err := w.engineClient.Log(w.ctx, &pulumirpc.LogRequest{
+		Message:   strings.ToValidUTF8(val, "�"),
+		Urn:       "",
+		Ephemeral: true,
+		StreamId:  w.streamID,
+		Severity:  w.severity,
+	}); err != nil {
+		return 0, err
+	}
+	return len(val), nil
+}
+
+// These packages are known not to have any plugins.
+var packagesWithoutPlugins = map[string]struct{}{
+	"pulumi-policy": {},
+}
+
+type pythonPackage struct {
+	Name    string `json:"name"`
+	Version string `json:"version"`
+}
+
+func determinePulumiPackages(virtualenv, cwd string) ([]pythonPackage, error) {
+	logging.V(5).Infof("GetRequiredPlugins: Determining pulumi packages")
+
+	// Run the `python -m pip list --format json` command.
+	args := []string{"-m", "pip", "list", "--format", "json"}
+	output, err := runPythonCommand(virtualenv, cwd, args...)
+	if err != nil {
+		return nil, err
+	}
+
+	// Parse the JSON output.
+	var packages []pythonPackage
+	if err := json.Unmarshal(output, &packages); err != nil {
+		return nil, errors.Wrapf(err, "parsing `python %s` output", strings.Join(args, " "))
+	}
+
+	// Only return Pulumi packages.
+	var pulumiPackages []pythonPackage
+	for _, pkg := range packages {
+		// We're only interested in packages that start with "pulumi-".
+		if !strings.HasPrefix(pkg.Name, "pulumi-") {
+			continue
+		}
```

Comment:
This is temporary - right?  In principle, we will want to support 3rd party packages with plugins?

Reply:
Yes, temporary. Go and .NET have this same limitation:

https://github.com/pulumi/pulumi/blob/002a228000bdcb10cc9851cc8bce09858065058b/sdk/go/pulumi-language-go/main.go#L140

https://github.com/pulumi/pulumi/blob/002a228000bdcb10cc9851cc8bce09858065058b/sdk/dotnet/cmd/pulumi-language-dotnet/main.go#L251-L252

As we work on making it possible/easier to publish 3rd party plugins, we'll need to address this across all the languages.